hathor/p2p/manager.py

```diff
@@ -581,93 +553,68 @@ def connect_to_if_not_connected(self, peer: UnverifiedPeer | PublicPeer, now: in
             # It makes no sense to keep storing peers that have disconnected and have no entrypoints
             # We will never be able to connect to them anymore and they will only keep spending memory
             # and other resources when used in APIs, so we are removing them here
-            if peer.id not in self.connected_peers:
+            if not self.is_peer_ready(peer.id):
                 self.verified_peer_storage.remove(peer)
             return
-        if peer.id in self.connected_peers:
+        if self.is_peer_ready(peer.id):
             return
 
         assert peer.id is not None
         if peer.info.can_retry(now):
             addr = self.rng.choice(peer.info.entrypoints)
             self.connect_to(addr.with_id(peer.id), peer)
 
-    def _connect_to_callback(
-        self,
-        protocol: IProtocol,
-        peer: UnverifiedPeer | PublicPeer | None,
-        endpoint: IStreamClientEndpoint,
-        entrypoint: PeerEndpoint,
-    ) -> None:
-        """Called when we successfully connect to a peer."""
-        if isinstance(protocol, HathorProtocol):
-            protocol.on_outbound_connect(entrypoint, peer)
-        else:
-            assert isinstance(protocol, TLSMemoryBIOProtocol)
-            assert isinstance(protocol.wrappedProtocol, HathorProtocol)
-            protocol.wrappedProtocol.on_outbound_connect(entrypoint, peer)
-        self.connecting_peers.pop(endpoint)
+    def _connect_to_callback(self, protocol: IProtocol, addr: PeerAddress, peer_id: PeerId | None) -> None:
+        """Called when we successfully connect to an outbound peer."""
+        if isinstance(protocol, TLSMemoryBIOProtocol):
+            protocol = protocol.wrappedProtocol
+        assert isinstance(protocol, HathorProtocol)
+        assert protocol.addr == addr
+        protocol.on_outbound_connect(peer_id)
 
     def connect_to(
         self,
-        entrypoint: PeerEndpoint,
+        endpoint: PeerEndpoint,
         peer: UnverifiedPeer | PublicPeer | None = None,
-        use_ssl: bool | None = None,
-    ) -> None:
-        """ Attempt to connect to a peer, even if a connection already exists.
-        Usually you should call `connect_to_if_not_connected`.
+    ) -> Deferred[IProtocol] | None:
+        """Attempt to connect to a peer though a specific endpoint."""
+        if endpoint.peer_id is not None and peer is not None:
+            assert endpoint.peer_id == peer.id, 'the entrypoint peer_id does not match the actual peer_id'
```

Comment:
This is an assertion because the only caller that passes a `peer` constructs the `endpoint` using the `peer.id`. In another PR this may be further improved so this assertion can be removed.